Add Octane GitHub guide. (#1832)
simplabs/ember-simple-auth: guides/octane-auth-torii-with-github.md

   GitHub account from your app. The user grants this.
2. GitHub redirects back to your app with an `authorizationCode`. This is
   proof of authorization, but your app still doesn’t know _who_ the user is
   (so it cannot connect to a `User` model, for example).
3. Your app sends the `authorizationCode` to a gatekeeper server you maintain,
   which in turn `POST`s the code to
   `https://github.com/login/oauth/access_token`. In development, this can be
   mocked. In production, it must be a separate server, about which more
   below, as we will use a separate server throughout.
4. GitHub responds to the `POST` with an `access_token`, which your gatekeeper
   server forwards to your app.
5. Your app uses the `access_token` in an `Authorization: Bearer
   <access_token>` header request to, say, `https://api.github.com/user`,
   which returns the GitHub user’s user info, at which point you can match the
   username with your `User` model or whatever. 

As with the original `ember-simple-auth` guide, the approach here mirrors any
OAuth2 explicit grant flow used also by Facebook, Google, etc. 

You can find a sequence diagram for the flow

> [here](http://github.com/simplabs/ember-simple-auth/guides/assets/esa-initial-flow.txt.pn://github.com/simplabs/ember-simple-auth/blob/master/guides/assets/esa-initial-flow.txt.png).

Reviewer: It seems the link is not pointing to the image yet - do you intend to do the following?

```suggestion
[here](https://github.com/simplabs/ember-simple-auth/blob/master/guides/assets/esa-initial-flow.txt.png).
```

Author: yes thank you.

## Set up and Configuration

The following guide will teach you about the basics of OAuth, Github and Ember, following these five major steps to implement Github Authorisation in your Ember app:

> experts, the short version is:

Reviewer: ```suggestion
experts, the short version is:
```

This seems to be just a remainder of the previous paragraph that had been edited

Author: Ah I misunderstood your review's scope. Yes, I'll rewrite that now.

Author: Rather, you're right… sloppy edit. Thanks!

   GitHub presents a new window asking the user to authorize access to their
   GitHub account from your app. The user grants this.
2. GitHub redirects back to your app with an `authorizationCode`. This is
   proof of authorization, but your app still doesn’t know _who_ the user is
   (so it cannot connect to a `User` model, for example).
3. Your app sends the `authorizationCode` to a gatekeeper server you maintain,
   which in turn `POST`s the code to
   `https://github.com/login/oauth/access_token`. In development, this can be
   mocked. In production, it must be a separate server, about which more
   below, as we will use a separate server throughout.
4. GitHub responds to the `POST` with an `access_token`, which your gatekeeper
   server forwards to your app.
5. Your app uses the `access_token` in an `Authorization: Bearer
   <access_token>` header request to, say, `https://api.github.com/user`,
   which returns the GitHub user’s user info, at which point you can match the
   username with your `User` model or whatever. 

As with the original `ember-simple-auth` guide, the approach here mirrors any
OAuth2 explicit grant flow used also by Facebook, Google, etc. 

> You can find a sequence diagram for the flow [here](http://github.com/simplabs/ember-simple-auth/guides/assets/esa-initial-flow.txt.png).

Reviewer: It seems the link does not return an image yet - can you check if this works for you?

Author: done.

## Set up and Configuration

> This section of the guide is for newcomers to OAuth, GitHub, and Ember. For

Reviewer: Personally, I think it would be nice to not create an arbitrary separation of the audience between "newbies who are in need of this guide"  and "experts who can skip it" and label them as "newbies" or "experts" depending on the fact if they find the TL;DR that is provided below clear already or not

I think people who read the TL;DR might in fact already be experts at Ember, OAuth and Github and still have more questions about the details of the setup which are explained in the extended guide. I think the TL;DR itself is super useful in any case, so what do you think about a more neutral intro along the lines of

```suggestion
The following guide will teach you about the basics of OAuth, Github and Ember, following these five major steps to implement Github Authorisation in your Ember app:
```

What are your thoughts?

Torii has not yet been configured, so let’s do that.

### Configure Torii

Torii is configured in the file `config/environment.js`. Somewhere around the
seventh line, perhaps underneath `rootURL: "/"`, add:

```javascript
    torii: {
      sessionServiceName: "session",
      providers: {
        "github-oauth2": {
          scope: "repo user",
          apiKey: process.env.GITHUB_DEV_CLIENT_ID,
          redirectUri: process.env.GITHUB_DEV_REDIRECT_URI,
          tokenExchangeUri: process.env.DEV_TOKEN_EXCHANGE_URL
        }
      }
    },
```

When you spin up the server again, it will not complain about Torii’s being
unconfigured.

## Implement the first stage of GitHub authorization

OAuth is officially an authorization protocol, but is commonly used also for
authentication when the initial authorization code is obtained over `https`.
GitHub uses the OAuth [authorization code grant
type](https://tools.ietf.org/html/rfc6749#section-4.1), which requires two
steps. As noted at the top, the first step uses your client ID to get a
temporary authorization code. The temporary authorization code acts as a
single use bridge to authorization. 

> ### Add `ember-simple-auth` Mixins to Routes

Reviewer 2: Any chance you could add how to use a custom mixin ... I'm trying to override `authenticationRoute`

popup window should appear asking for authorization, showing a GitHub icon and
your GitHub user avatar. Upon clicking the green “Authorize” button, the popup
should disappear and dump you back to your app, but now on the `index` route.
Underneath “Index Route,” you should also now see a line displaying the GitHub
authorization code.

We’ve now established the mechanism to obtain an authorization code from
GitHub. This doesn't authorize us fully to use the GitHub APIs, although
`ember-simple-auth` considers us authenticated at this point, hence
`this.session.isAuthenticated` evaluates to `true`.

The “Logout” button in the `application` template now appears, and if you
click on it, you will be dumped back to the `login` route. Logging in,
however, no longer requires authorization via the GitHub popup, because you
have already authorized your app. The popup still appears, but it flashes for
a second and then disappears.

This completes the second step of the outline at the top of this guide. Next
is performing the token exchange to enable getting user data from GitHub.

## Deploy a Gatekeeper Server for Token Exchange and Connect It to Ember

> Here the app has to move out of Emberdom for a bit, because GitHub will not

Reviewer: I'm wondering what is meant by `Emberdom` - can you outline it?

Author: I've clarified the wording. I just meant that the next step happens outside of the Ember app. This was, conceptually, the hardest part for me to understand.

    }

    return headers;
  })
});
```

This adapter injects an authorization header into the GitHub request now.  The
`DataAdapterMixin` is a mixin provided by `ember-simple-auth` that injects the
`session` service. Otherwise, when you start up your server one last time, log
in via the `login` route, and get dumped back to the `index` route, you should
now see your avatar, your GitHub login name, and your full name.

Congrats!

## Next Steps

This all done, you have a crude Ember Octane app that authenticates with
GitHub, returning the logged in user. This model can then be used to populate
your “real” `User` model, perhaps.

Additionally, in production you cannot rely on using your local Gatekeeper
server, but luckily Gatekeeper includes documentation on [deploying to
Heroku](https://github.com/prose/gatekeeper#deploy-on-heroku) and similar
services. Then, update the appropriate line in your Ember app’s `.env` to
point to the Heroku url, not `http://localhost:9999/authenticate`.

Finally, I’ve created a repo of the “finished” version of this guide, and it
is available at
[@muziejus/ember-simple-auth-github-octane](http://github.com/muziejus/ember-simple-auth-github-octane).

> Here are some other links provided by [@srvance](http://github.com/srvance),

Reviewer: It's really cool that you added further resources for learning and the related credit here :+1:
